File: haskell/bob/bob.hs
module Bob where

import Data.Char

responseFor :: String -> String
responseFor stmt 
    | silent    = "Fine. Be that way!"
    | shout     = "Whoa, chill out!"
    | question  = "Sure."
    | otherwise = "Whatever."
        where silent   = all isSpace stmt
              question = last stmt == '?'
              shout    = all isUpper letters && not (null letters) 
              letters  = filter isAlpha stmt
    

File: clojure/bob/bob.clj
(ns bob)

(defn silence? [phrase] 
  (clojure.string/blank? phrase))

(defn shout? [phrase] 
  (and
    (= phrase (clojure.string/upper-case phrase))
    (not (nil? (re-find #"[A-Z]" phrase)))))

(defn question? [phrase] 
  (not (nil? (re-find #"\?$" phrase))))

(defn response-for [phrase]
    (cond
      (silence? phrase) "Fine. Be that way!"
      (shout? phrase) "Whoa, chill out!"
      (question? phrase) "Sure."
      :else "Whatever."))
